addons/@mlabs-haskell/godot-cardano/src/single_address_wallet_store.gd

```diff
@@ -0,0 +1,140 @@
+extends Node
+
+class_name SingleAddressWalletLoader
+
+enum Status {
+	SUCCESS = 0,
+	BAD_PHRASE = 1,
+	BIP32_ERROR = 2,
+	PKCS5_ERROR = 3,
+	BAD_SCRYPT_PARAMS = 4,
+	COULD_NOT_PARSE_AES_IV = 5,
+	ACCOUNT_NOT_FOUND = 6
+}
+
+## Emitted when [method SingleAddressWalletStore.import_from_seedphrase]
+## returns a result.
+signal import_completed(res: WalletImportResult)
+
+## May be null if no wallet was loaded
+var _wallet_store : _SingleAddressWalletStore
+
+# Used when importing
+var thread: Thread
+
+func _init() -> void:
+	pass
+
+class GetWalletError extends Result:
+	## WARNING: This function may fail! First match on [Result_.tag] or call [Result_.is_ok].
+	var value: SingleAddressWallet:
+		get: return SingleAddressWallet.new(_res.unsafe_value() as _SingleAddressWallet)
+	## WARNING: This function may fail! First match on [Result_.tag] or call [Result._is_err].
+	var error: String:
+		get: return _res.unsafe_error()
+		
+class WalletImport extends RefCounted:
+	var _import_res: _SingleAddressWalletImportResult
+	var wallet: SingleAddressWallet:
+		get: return SingleAddressWallet.new(_import_res.wallet)
+	
+	func _init(import_res: _SingleAddressWalletImportResult):
+		_import_res = import_res
+
+class WalletImportResult extends Result:
+	## WARNING: This function may fail! First match on [Result_.tag] or call [Result_.is_ok].
+	var value: WalletImport:
+		get: return WalletImport.new(_res.unsafe_value() as _SingleAddressWalletImportResult)
+	## WARNING: This function may fail! First match on [Result_.tag] or call [Result._is_err].
+	var error: String:
+		get: return _res.unsafe_error()
+
+## Construct a [SingleAddressWalletStoreError] from a mnemonic [param phrase].
+## 
+## The phrase should follow the BIP32 standard and it may have a [phrase_password]
+## (if not, an empty string should be passed). This standard is followed by
+## almost all Cardano light wallets, so importing an already existing wallet
+## is normally achieved by calling this method with the wallet's seed phrase
+## and an empty password.
+##
+## This function also assigns a [param wallet_password] to the wallet that must be
+## entered for any method that requires it (such as signing or adding new accounts).
+##
+## Optionally, a [param account_index], [param name] and
+## [param account_description] may be passed. By default, the account index is
+## set to 0, with all subsequent accounts added to the wallet taking the next
+## indices.
+##
+## This function is asynchronous. You can await it or use hook a callback to
+## [signal SingleAddressWalletStore.import_completed].
+func import_from_seedphrase(
+	phrase: String,
+	phrase_password: String, 
+	wallet_password: String,
+	account_index: int,
+	name: String,
+	account_description: String) -> WalletImportResult:
+		if not (thread == null) and thread.is_alive():
+			push_warning("Import in progress, ignoring latest call...")
+			await(import_completed)
+		
+		thread = Thread.new()
```

Comment:
I see! There is an issue, albeit a different one in my opinion. I intended to await for the existing import to finish and then return the result, but I forgot the last part. That's why it looks non-sensical now.

The issue with waiting for the thread to finish should be handled by the callback `_exit_tree`, which should always wait for the thread if it is alive.

Reply:
@danielfarrelly Will push a commit with a fix for this. It will essentially wait for the thread to finish _only_ if it is not longer alive (which should be instantaneous). Otherwise, it will await for the `import_completed` and return the old result.